pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/immutable/ImmutableSegmentImpl.java

```diff
@@ -283,31 +276,22 @@ public ThreadSafeMutableRoaringBitmap getValidDocIds() {
 
   @Override
   public GenericRow getRecord(int docId, GenericRow reuse) {
-    try {
-      if (_pinotSegmentRecordReader == null) {
-        _pinotSegmentRecordReader = new PinotSegmentRecordReader();
-        _pinotSegmentRecordReader.init(this);
-      }
-      _pinotSegmentRecordReader.getRecord(reuse, docId);
+    try (PinotSegmentRecordReader recordReader = new PinotSegmentRecordReader()) {
```

Comment:
That's why we should avoid using it unless unavoidable (partial-upsert). When reading multiple records, we should create the record reader instead of using this method (mentioned in the javadoc).

Reply:
But partial upsert is used by a lot of users. can we instead implement another method where we can pass a segment reader?
then we can simply init and close that reader outside the segment

Reply:
We can directly read records from segment reader, so no need to add another method into the segment. The problem for partial upsert is that we cannot keep the segment reader open, or it will consume too much memory. As mentioned in the description, since partial-upsert update is not on query path, we can merge it for now, and if people reporting noticeable regression, we can consider not allowing compression when partial-upsert is enabled.